### DISCUSSION TITLE
The two programs call each other, showing a hierarchical error

### DISCUSSION THREAD

**Discussion by Mrwu555** [kimmknight/remoteapptool#113] [Q&A]
I am hosting my remote apps on Windows Server 2022 Standard(64bit).
When I connect to an app from a client PC,the app launches as expected.
My program is written in VB.NET language, the composition of the program is A.exe calls the B.exe,
the A.exe is composed of one form, the B.exe is composed of two forms, they are parent-child relationships,
in the process of use, we found that when the B.exe child form is closed, 
the A.exe form will be called to the front, this is not the desired behavior, 
the correct situation is to call the parent form of the B.exe to the front.

Is the behavior that I am observing the expected behavior, or is it due to a misconfiguration on my end?

**Comment by MrBrianGale** [comment]
My opinion, this sounds like an application development related bug not a bug with remoteapptool.
Do you get the same behavior when running the apps outside of a remoteapp session? I have a strong suspicion that you will...

Also, just my 2 cents, but that is not the most sustainable way to design your application to begin with. IF application B relies on application A to call it, why not build application B INTO application A? The only exception I can personally think of to this is if application A is more of a "launcher" tool that can start up multiple different apps, but if that was the case, I would not have a remoteapp session to start an app launcher. I would have the app launcher run on the local machines and call the remoteapp instead. The main reason being that end user resources are cheap and plentiful. Server resources are expensive and shared.
BUT if A.exe is a standalone application that does something then calls B.exe (such as A.exe handles login and database connection string selection and B.exe is the "main" application), I would just bake B.exe into A.exe. As for B.exe child form close, it really depends on how you designed the app. It MAY be that when you close the child form, it is calling exit operation on the application level, not on the form level. If this is the case, closing B.exe's child form will close B.exe's parent form too. The other thing you could do is in the B.exe child form is make a call to Windows to bring B.exe's main window to focus and bring it on top of other windows so it comes to the front.

Now, IF this is a bug in remoteapp, it is still not a bug in remoteapptool. RemoteAppTool is just a front-end for making the .RDP file and doing some back-end tweaks server side so that the app is presented as a remote app. IF the behavior only happens in remote apps, then the problem is with your app not being remote app aware OR the remote app tool you are using  (msdtc.exe for example on windows) misinterpreting a form close event and doing a request to change the application focus.

The TL;DR version - this sounds like an application bug in B.exe and/or A.exe, not a bug in remoteapptool. Worst case, it's a bug in the remote app tool you are using and you'll need to open a support request with the app vendor. Best case, it is a bug in the app you developed (A.exe and/or B.exe) and you can modify the source to make sure that the application focus happens in the way you expect.

Does the above help?

**Reply by Mrwu555** [reply]
Thank you very much for your answer.
There is no same behavior when running an app outside of a remoteapp session.
And for some reason, the structure of the program cannot be modified, and it can only be kept as it is.
I am hosting my remote apps on Windows 11 Professional 22H2(64bit).The same problem will still occur.
Is there any other solution?

**Reply by Mrwu555** [reply]
I made another example, A.exe called B.exe, form1 in the B.exe, and called Form2 in the form showDialog.When you close the B.exe Form2 and then close the B.exe Form1, the A.exe Form will be hidden to the back of the application.
There is no same behavior when running an app outside of a remoteapp session.
Is the behavior that I am observing the expected behavior, or is it due to a misconfiguration on my end?
